File: sw/qa/complex/writer/CheckIndex.java
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package complex.writer;

import com.sun.star.uno.UnoRuntime;
import com.sun.star.uno.XComponentContext;
import com.sun.star.lang.EventObject;
import com.sun.star.lang.XMultiServiceFactory;
import com.sun.star.beans.XPropertySet;
import com.sun.star.util.XRefreshable;
import com.sun.star.util.XRefreshListener;
import com.sun.star.text.ControlCharacter;
import com.sun.star.text.XDocumentIndex;
import com.sun.star.text.XParagraphCursor;
import com.sun.star.text.XText;
import com.sun.star.text.XTextContent;
import com.sun.star.text.XTextDocument;
import org.openoffice.test.OfficeConnection;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import static org.junit.Assert.*;


public class CheckIndex
{
    private static final OfficeConnection connection = new OfficeConnection();

    @BeforeClass public static void setUpConnection() throws Exception {
        connection.setUp();
    }

    @AfterClass public static void tearDownConnection()
        throws InterruptedException, com.sun.star.uno.Exception
    {
        connection.tearDown();
    }

    private XMultiServiceFactory m_xMSF = null;
    private XComponentContext m_xContext = null;
    private XTextDocument m_xDoc = null;

    @Before public void before() throws Exception
    {
        m_xMSF = UnoRuntime.queryInterface(
            XMultiServiceFactory.class,
            connection.getComponentContext().getServiceManager());
        m_xContext = connection.getComponentContext();
        assertNotNull("could not get component context.", m_xContext);
        m_xDoc = util.WriterTools.createTextDoc(m_xMSF);
    }

    @After public void after()
    {
        util.DesktopTools.closeDoc(m_xDoc);
    }

    class RefreshListener implements XRefreshListener
    {
        public boolean m_bDisposed = false;
        public boolean m_bRefreshed = false;
        public void disposing(EventObject event)
        {
            m_bDisposed = true;
        }
        public void refreshed(EventObject event)
        {
            m_bRefreshed = true;
        }
        public void assertRefreshed()
        {
            assertTrue(m_bRefreshed);
            m_bRefreshed = false;
        }
    }

    @Test
    public void test_refresh() throws Exception
    {
        XMultiServiceFactory xDocFactory =
            UnoRuntime.queryInterface(XMultiServiceFactory.class, m_xDoc);
        Object xIndex =
            xDocFactory.createInstance("com.sun.star.text.ContentIndex");

        XText xBodyText = m_xDoc.getText();
        XParagraphCursor xCursor = UnoRuntime.queryInterface(
                XParagraphCursor.class, xBodyText.createTextCursor());
        XPropertySet xCursorSet =
            UnoRuntime.queryInterface(XPropertySet.class, xCursor);
        XTextContent xIndexContent =
            UnoRuntime.queryInterface(XTextContent.class, xIndex);
        XPropertySet xIndexSet =
            UnoRuntime.queryInterface(XPropertySet.class, xIndex);
        xIndexSet.setPropertyValue("CreateFromOutline", true);
        xBodyText.insertTextContent(xCursor, xIndexContent, true);

        XRefreshable xRefreshable =
            UnoRuntime.queryInterface(XRefreshable.class, xIndex);

        // test that refresh calls listener
        RefreshListener listener = new RefreshListener();
        xRefreshable.addRefreshListener(listener);
        assertFalse(listener.m_bRefreshed);
        xRefreshable.refresh();
        listener.assertRefreshed();

        // insert some heading
        xCursor.gotoEnd(false);
        xBodyText.insertControlCharacter(xCursor,
                ControlCharacter.PARAGRAPH_BREAK, false);
        xCursor.gotoEnd(false);
        xCursor.setString("a heading");
        xCursor.gotoStartOfParagraph(true);
        xCursorSet.setPropertyValue("ParaStyleName", "Heading 1");

        xRefreshable.refresh();
        listener.assertRefreshed();
        // hope text is in last paragraph...
        xCursor.gotoRange(xIndexContent.getAnchor().getEnd(), false);
        xCursor.gotoStartOfParagraph(true);
        String text = xCursor.getString();
        assertTrue(text.contains("a heading"));

        // insert some more headings
        xCursor.gotoEnd(false);
        xBodyText.insertControlCharacter(xCursor,
                ControlCharacter.PARAGRAPH_BREAK, false);
        xCursor.gotoEnd(false);
        xCursor.setString("yet another heading");
        xCursor.gotoStartOfParagraph(true);
        xCursorSet.setPropertyValue("ParaStyleName", "Heading 1");

        // try again with update
        XDocumentIndex xIndexIndex =
            UnoRuntime.queryInterface(XDocumentIndex.class, xIndex);
        xIndexIndex.update();
        listener.assertRefreshed();
        xCursor.gotoRange(xIndexContent.getAnchor().getEnd(), false);
        xCursor.gotoStartOfParagraph(true);
        text = xCursor.getString();
        assertTrue(text.contains("yet another heading"));

        // dispose must call the listener
        assertFalse(listener.m_bDisposed);
        xIndexIndex.dispose();
        assertTrue(listener.m_bDisposed);
    }
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: sw/qa/python/check_index.py
#! /usr/bin/env python
# -*- tab-width: 4; indent-tabs-mode: nil; py-indent-offset: 4 -*-
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
import unittest
import unohelper
from org.libreoffice.unotest import UnoInProcess
from com.sun.star.text.ControlCharacter import PARAGRAPH_BREAK
from com.sun.star.util import XRefreshListener


class RefreshListener(XRefreshListener, unohelper.Base):
    def __init__(self):
        self.m_refreshed = False
        self.m_disposed = False

    # Gets called when index is disposed
    def disposing(self, event):
        self.m_disposed = True

    # Gets called when index is refreshed
    def refreshed(self, event):
        self.m_refreshed = True


class CheckIndex(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls._uno = UnoInProcess()
        cls._uno.setUp()

    @classmethod
    def tearDownClass(cls):
        cls._uno.tearDown()

    # Gets called every time a new test is run
    def setUp(self):
        """
        Every test should run with a new Instance of an Index and
        Refresh Listener
        So we need to reload the text of the document and initialize
        the corresponding Cursor
        """
        self.xDoc = self.__class__._uno.openEmptyWriterDoc()
        self.xIndex = self.xDoc.createInstance(
            "com.sun.star.text.ContentIndex")
        self.xBodyText = self.xDoc.getText()
        self.xCursor = self.xBodyText.createTextCursor()
        self.xIndex.setPropertyValue("CreateFromOutline", True)
        self.xBodyText.insertTextContent(self.xCursor, self.xIndex, True)
        # Create Refresh Listener and bind it to the Index
        self.listener = RefreshListener()
        self.xIndex.addRefreshListener(self.listener)

    def tearDown(self):
        """
        Dispose Index and Document and check if the index was
        deleted while the tests
        """
        self.assertFalse(self.listener.m_disposed,
                         "Unexpected disparue of the Refresh Listener!")
        self.xIndex.dispose()
        self.assertTrue(self.listener.m_disposed,
                        "Could not dispose Refresh Listener")
        self.xDoc.dispose()

    def insert_heading(self, text):
        """
        Insert a Heading at the end of the document
        """
        self.xCursor.gotoEnd(False)
        self.xBodyText.insertControlCharacter(self.xCursor,
                                              PARAGRAPH_BREAK, False)
        self.xCursor.gotoEnd(False)
        self.xCursor.setString(text)
        self.xCursor.gotoStartOfParagraph(True)
        self.xCursor.setPropertyValue("ParaStyleName", "Heading 1")

    def test_index_refresh(self):
        """
        Try to insert a heading at the index, refresh the index and
        retrieve the heading again
        """
        heading = "The best test heading you have seen in your entire life"
        self.insert_heading(heading)
        self.xIndex.refresh()
        self.assertTrue(self.listener.m_refreshed, "Failed to refresh index!")
        self.listener.m_refreshed = False
        self.xCursor.gotoRange(self.xIndex.getAnchor().getEnd(), False)
        self.xCursor.gotoStartOfParagraph(True)
        # Get String at current position and search for the heading
        text = self.xCursor.getString()
        self.assertGreaterEqual(text.find(heading), 0,
                                "Failed to insert heading at index "
                                "and retrieve it again!")

    def test_index_update(self):
        """
        Try to insert a heading at the index, update the index
        and retrieve the heading again
        """
        heading = "Heading to test the index update"
        self.insert_heading(heading)
        self.xIndex.update()
        self.assertTrue(self.listener.m_refreshed, "Failed to update index!")
        self.listener.m_refreshed = False
        self.xCursor.gotoRange(self.xIndex.getAnchor().getEnd(), False)
        self.xCursor.gotoStartOfParagraph(True)
        # Get String at current position and search for the heading
        text = self.xCursor.getString()
        self.assertGreaterEqual(text.find(heading), 0,
                                "Failed to insert a heading at Index and "
                                "retrieve it again!")


if __name__ == "__main__":
    unittest.main()
